#!/bin/bash

# This is MAP.pbs
# 
#PBS -l nodes=1:ppn=1:thinnode
#PBS -l walltime=2:00:00:00
#PBS -l mem=1gb

# Example usage:
# myqsub -d `pwd` -t 1 -F "${ARGS}" phamb_pipe/COVERAGE.pbs

### exit if any errors or unset variables are encountered
#set -euo pipefail

### LOAD MODULES NEEDED
module load tools 
module load perl/5.24.0 
module load samtools/1.9
module load bedtools/2.28.0

### echo commands to the error stream (may help debugging)
set -x 

### GLOBAL VARIABLES
readonly INPUT_TABLE="$1"
readonly OUTPUTDIR="$2"
readonly THREADS=4

### echo commands to the error stream (may help debugging)
set -x 
shift
shift
shift
shift
shift
shift
readonly LOGDIR=log
readonly JOBID_SIMPLE=`echo ${PBS_JOBID} | sed -r 's/(\[|\.).*$//'`   ## extract the numeric part only
readonly TEMPDIR=tmp/mapping

## read from a single line corresponding to the job array ID
read SAMPLE FQ1 FQ2 PLACEHOLDER < <(cat $INPUT_TABLE | sed "${PBS_ARRAYID}q;d")

### Samtools from v. 1.9 sorts with -m 768M  , so using 1 gb should be possible?

if [ ! -f "${OUTPUTDIR}/${SAMPLE}.position.coverage" ]; then
samtools sort --threads ${THREADS} -o ${OUTPUTDIR}/${SAMPLE}.position.bam -O BAM ${OUTPUTDIR}/${SAMPLE}.bam
bedtools genomecov -ibam ${OUTPUTDIR}/${SAMPLE}.position.bam > ${OUTPUTDIR}/${SAMPLE}.position.coverage
fi

### Write out the fraction of the contig with Depth Zero 
awk '$2==0' ${OUTPUTDIR}/${SAMPLE}.position.coverage > ${OUTPUTDIR}/${SAMPLE}.position.zero.coverage